### PR TITLE
add deterministic embedding grad kernel

### PR DESCRIPTION
**Pull request by Wong4j**
### PR types

Performance optimization
### PR changes

OPs
### Description

Issue: 
Embedding backward kernel is non-deterministic (due to atomicAdd) by default. If users set "FLAGS_cudnn_deterministic=1",  threadsPerBlock will be set to (128, 1), num blocks will be set to 1, which ensures determinism but is very slow.

Solution: 
I add a deterministic embedding backward kernel (refer to PyTorch's [implementation](https://github.com/pytorch/pytorch/blob/28ed0bdb3701cdf71b2e281a4c00f7318db8c98a/aten/src/ATen/native/cuda/Embedding.cu#L94)). It's slower than the current non-deterministic implementation but much faster than the current deterministic method.

Perf tests:
test case - BERT large phase1. (including 3 embedding layer)
overall training speed: 415 ms/step

unit: ms
image

Obviously, paddle's current deterministic implementation is quite slow, which will significantly affect the training speed of the BERT model.

### CODE REVIEW

File: paddle/phi/core/flags.cc [hunk 1]
```diff
@@ -264,6 +264,22 @@ PADDLE_DEFINE_EXPORTED_bool(
     "operator. The autotuning algorithm may be non-deterministic. If "
     "true, the algorithm is deterministic.");
 
+/**
+ * CUDA related FLAG
+ * Name: FLAGS_embedding_deterministic
+ * Since Version: 2.4
```
Comment (Xreki) [inline paddle/phi/core/flags.cc]: 开始支持的版本号请改成下一个版本，即2.5。

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 2]
```diff
@@ -104,13 +191,20 @@ struct EmbeddingGradCUDAFunctor {
       dim3 threads(128, 8);
       dim3 grids(gridx, 1);
 
-      if (FLAGS_cudnn_deterministic) {
-        VLOG(2) << "Run grad kernel of embedding with single thread.";
-        grids.x = 1;
-        threads.y = 1;
+      dim3 threads2(WARP_SIZE, BLOCKDIMY);
```
Comment (shaojiewang) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: [not necessary] Variable ```threads2``` and ```grid2``` 's naming style is not very clear. ```threads_deterministic``` could be a choice. As their scope is only in the statement of ```if (FLAGS_cudnn_deterministic || FLAGS_embedding_deterministic)```, they could be moved into this statement.

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 3]
```diff
@@ -61,6 +72,82 @@ __global__ void EmbeddingGrad(T* table,
...
+__global__ void EmbeddingGradDeterministic(T* table,
+                                           const T* output,
+                                           const IdT* ids,
+                                           const int64_t K,
+                                           const int64_t D) {
+  extern __shared__ char buf[];
+  T* smem = reinterpret_cast<T*>(buf);
+  T* my_s = smem + WARP_SIZE * threadIdx.y;
+  int* indices_batch =
+      reinterpret_cast<int*>(buf + sizeof(T) * WARP_SIZE * blockDim.y);
```
Comment (shaojiewang) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: If ```blockDim.y ``` is forced to be ```BLOCKDIMY```, ```blockDim.y``` in kernel could be replaced by ```BLOCKDIMY```.

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 4]
```diff
@@ -104,13 +191,20 @@ struct EmbeddingGradCUDAFunctor {
...
       dim3 grids(gridx, 1);
 
-      if (FLAGS_cudnn_deterministic) {
-        VLOG(2) << "Run grad kernel of embedding with single thread.";
-        grids.x = 1;
-        threads.y = 1;
+      dim3 threads2(WARP_SIZE, BLOCKDIMY);
+      dim3 grids2(static_cast<int>((D + WARP_SIZE - 1) / WARP_SIZE));
+
+      if (FLAGS_cudnn_deterministic || FLAGS_embedding_deterministic) {
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: 若引入新的控制FLAGS `FLAGS_embedding_deterministic`，这里建议不再使用`FLAGS_cudnn_deterministic`。

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 5]
```diff
@@ -104,13 +191,20 @@ struct EmbeddingGradCUDAFunctor {
       dim3 threads(128, 8);
       dim3 grids(gridx, 1);
 
-      if (FLAGS_cudnn_deterministic) {
-        VLOG(2) << "Run grad kernel of embedding with single thread.";
-        grids.x = 1;
-        threads.y = 1;
+      dim3 threads2(WARP_SIZE, BLOCKDIMY);
+      dim3 grids2(static_cast<int>((D + WARP_SIZE - 1) / WARP_SIZE));
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: 请将L190-L192挪入`else`分支，L194-L195挪入`if`分支。

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 6]
```diff
@@ -61,6 +72,82 @@ __global__ void EmbeddingGrad(T* table,
   }
 }
 
+template <typename T, typename IdT>
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu:76]: 请添加一些注释，说明该函数实现的原理吧。

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 7]
```diff
@@ -61,6 +72,82 @@ __global__ void EmbeddingGrad(T* table,
...
+              (dst_row ==
+               indices_batch[chunk_start - batch_start + threadIdx.x]);
+#ifdef PADDLE_WITH_HIP
+        unsigned long long int matchmask =      // NOLINT
+            __ballot(match_found_this_thread);  // NOLINT
+        int first_remaining_peer = __ffsll(matchmask) - 1;
+#else
+        unsigned int matchmask = __ballot_sync(MASK, match_found_this_thread);
+        int first_remaining_peer = __ffs(matchmask) - 1;
+#endif
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu:139]: 请加一些注释。

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 8]
```diff
@@ -61,6 +72,82 @@ __global__ void EmbeddingGrad(T* table,
...
+      int src_row = chunk_start + threadIdx.y;
+      int dst_row = indices_batch[src_row - batch_start];
+      if (src_row < K && f < s)
+        my_s[threadIdx.x] = static_cast<T>(output[src_row * D + f]);
+
+      __syncthreads();
+
+      if (src_row < K) {
+        int match_found_this_thread = 0;
+        if (threadIdx.x < n_this_chunk)
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: `if`后添加`{}`，加强代码的可读性。

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 9]
```diff
@@ -61,6 +72,82 @@ __global__ void EmbeddingGrad(T* table,
...
+
+  const int f = threadIdx.x + blockIdx.x * blockDim.x;
+
+  for (int batch_start = 0; batch_start < K;
+       batch_start += blockDim.x * blockDim.y) {
+    int tid = threadIdx.x + threadIdx.y * blockDim.x;
+    if (batch_start + tid < K)
+      indices_batch[tid] = static_cast<int>(ids[batch_start + tid]);
+
+    int batch_end = batch_start + blockDim.x * blockDim.y < K
```
Comment (shaojiewang) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: ```min(batch_start + WARP_SIZE * BLOCKDIMY, K)```

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 10]
```diff
@@ -61,6 +72,82 @@ __global__ void EmbeddingGrad(T* table,
...
+
+  const int s = static_cast<int>(D);
+
+  const int f = threadIdx.x + blockIdx.x * blockDim.x;
+
+  for (int batch_start = 0; batch_start < K;
+       batch_start += blockDim.x * blockDim.y) {
+    int tid = threadIdx.x + threadIdx.y * blockDim.x;
+    if (batch_start + tid < K)
+      indices_batch[tid] = static_cast<int>(ids[batch_start + tid]);
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: `ids`的类型可能是`int64_t*`，这里强制转换成`int`类型，极端情况下可能会出错。

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 11]
```diff
@@ -61,6 +72,82 @@ __global__ void EmbeddingGrad(T* table,
...
+                                           const int64_t D) {
+  extern __shared__ char buf[];
+  T* smem = reinterpret_cast<T*>(buf);
+  T* my_s = smem + WARP_SIZE * threadIdx.y;
+  int* indices_batch =
+      reinterpret_cast<int*>(buf + sizeof(T) * WARP_SIZE * blockDim.y);
+
+  const int s = static_cast<int>(D);
+
+  const int f = threadIdx.x + blockIdx.x * blockDim.x;
```
Comment (shaojiewang) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: As per torch's code, ```s``` means stride and ```f``` means feature_dim. Could you please rename them or add comments here?

**Review by shaojiewang — CHANGES_REQUESTED** [review]
could you please add some comments in codes? the logic is not easy to understand.

File: paddle/phi/core/flags.cc (hunk 1 above)
Comment (Wong4j) [reply paddle/phi/core/flags.cc]: Done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 4 above)
Comment (Wong4j) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: Done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 5 above)
Comment (Wong4j) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: Done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 6 above)
Comment (Wong4j) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu:76]: 已经添加了注释

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 7 above)
Comment (Wong4j) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu:139]: Done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 8 above)
Comment (Wong4j) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: Done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 10 above)
Comment (Wong4j) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: Done.
一般id不会超过int的上限。我已经改成了int64_t，会消耗更多的shared memory，对性能影响不大

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 2 above)
Comment (Wong4j) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: Done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 3 above)
Comment (Wong4j) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: Done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 9 above)
Comment (Wong4j) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: Done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 11 above)
Comment (Wong4j) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: Done

**Review by Xreki — APPROVED** [review]
LGTM. 后续需要再提PR增加单测，保证计算的正确性。

**Review by shaojiewang — APPROVED** [review]

File: paddle/phi/core/flags.cc [hunk 12]
```diff
@@ -264,6 +264,22 @@ PADDLE_DEFINE_EXPORTED_bool(
...
+ * CUDA related FLAG
+ * Name: FLAGS_embedding_deterministic
+ * Since Version: 2.5
+ * Value Range: bool, default=false
+ * Example:
+ * Note: whether to use deterministic algorithm in embedding op.
+ *       If true, it will use deterministic CUDA kernel in embedding op.
+ */
+PADDLE_DEFINE_EXPORTED_bool(
+    embedding_deterministic,
```
Comment (phlrain) [inline paddle/phi/core/flags.cc:245]: 经过@Xreki 和 我的讨论， 我们建议把 embedding_deterministic 改成 paddel_deterministic，其他的算子统一用这个flags，麻烦看下这种改动是否能满足你们的需求

Comment (Wong4j) [reply paddle/phi/core/flags.cc:245]: 如果deterministic的功能仅仅用来debug，不考虑性能，你说的统一FLAG的设计是很好的。但是paddle中有其他的几个op是non-deterministic的，主要是因为用到了atomicAdd，比如[index_select_grad_kernel.cu#L95](https://github.com/PaddlePaddle/Paddle/blob/8e7c37894f8381d5e9662aa2974c4bbeac9b628e/paddle/phi/kernels/gpu/index_select_grad_kernel.cu#L95)。paddle目前的做法是启动一个block一个thread来跑，保证determinism，这样性能会极其的差。

我修改这个embedding_grad主要是想让BERT的训练结果是deterministic的。但是BERT中会用到`embedding_grad`和`index_select_grad`这两个non-deterministic的op。如果用同一个FLAG来控制所有op开启deterministic，`index_select_grad`会变得极其慢。所以我用单独的一个FLAG来开启deterministic的embedding_grad，以免影响BERT的训练速度。

实际上index_select这个op在BERT里面不会影响determinism，因为不会select重复的index，所以反向op中atomicAdd也不存在多threads累加到同一个地址的情况。所以我只需要deterministic的embedding就可以保证BERT的loss是完全稳定的。

Comment (Wong4j) [reply paddle/phi/core/flags.cc:245]: 如果你们讨论决定用统一的FLAG，我就按你们的意见来修改。
大多数用户没有我这个需求，所以添加单独的FLAG似乎不太好。我也考虑过类似于nn.embedding(arg1, arg2, ..., deterministic=True)这样的设计，这样也不好。
你们对此有什么意见吗？

**Review by phlrain — APPROVED** [review]

**Comment by Xreki** [comment]
CI无法rerun，请merge一下develop吧

**Comment by Xreki** [comment]
@Wong4j ，功能已合入，请尽快提PR添加单测。

**Pull request merged**